### ISSUE TITLE
fp16 inference is slower than fp32 on Nvidia Jetson TX2

### ISSUE THREAD

**Issue by adavoudi** [tensorflow/tensorflow#15643]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 16.04
- **TensorFlow installed from (source or binary)**: Source
- **TensorFlow version (use command below)**: r1.5
- **Python version**: 3.5.2
- **Bazel version (if compiling from source)**: 0.9.0
- **GCC/Compiler version (if compiling from source)**: 5.4.0
- **CUDA/cuDNN version**: Cuda 8.0 and cuDNN 7.0
- **GPU model and memory**: Jetson TX2

### The problem
I created a fully convolutional float 16 (half precision) neural network in tensorflow. When I run this network with some inputs, the inference time is slower than when I run the same network in float 32 (full precision) mode. 
I should also note that the following variables are set:

`
os.environ['TF_FP16_CONV_USE_FP32_COMPUTE'] = '0'
os.environ['TF_FP16_MATMUL_USE_FP32_COMPUTE'] = '0'
`

As Nvidia Jetson TX2 support FP16 operations, I expected an inference time not worse than when I use FP32, but surprisingly it is about 1.5 times worse! (36 miliseconds vs 22 miliseconds). I guess it is becuase of the overhead of internal type conversion in the tensorflow core between float16 and float32!

Is it a problem with Tensorflow or TX2?

**Comment by bignamehyp** [comment]
@yzhwang any idea?

**Comment by yzhwang** [comment]
Thank you @alirezadavoudi ! We are aware of this and we have observed this on P100. We are communicating with NVIDIA with solution for this.

First of all, could you try with CUDA 9 + cudnn 7 and see if the same result still shows up?

Second, in your setting, you are forcing cudnn to use TRUE_HALF_CONFIG and this mode is only supported by one internal kernel: CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_PRECOMP_GEMM as for now.  Sometimes for the same input shape, switching from FP32 to FP16 will cause the change of the tile size for this kernel. For example, an original 128*32 tile size could become 256*64 when switching from FP32 to FP16, while in this case, the gain from doubling the math is lost for the tile utilization.

In the future we will make PSEUDO_HALF_CONFIG and TRUE_HALF_CONFIG transparent to the users by adding these two configs to autotune.

**Comment by alirezadavoudi** [comment]
@yzhwang, unfortunately, CUDA 9 is still not released for TX2. 
It seems right now the best practice is to use TensorRT. But it would be great if Tensorflow could support native FP16.

**Comment by tokk-nv** [comment]
CUDA 9.0 is actually available in JetPack 3.2 Developer Preview.
https://developer.nvidia.com/embedded/jetpack-notes

**Comment by KillPinguin** [comment]
@alirezadavoudi 
I'm not experienced with Jetson, but would it be possible to profile your application with nvprof? In #15585 I described some limiting behavior with FP16, maybe this is an issue for you, too?

Would love to hear your thoughts on this, since it seems that we have similar performance issues.

**Comment by tensorflowbutler** [comment]
A member of the TensorFlow organization has replied after the stat:awaiting tensorflower label was applied.

**Comment by reedwm** [comment]
This comment from @yzhwang seems to be the main issue (although it's unknown if this will significantly increase performance):
>In the future we will make PSEUDO_HALF_CONFIG and TRUE_HALF_CONFIG transparent to the users by adding these two configs to autotune.

So, I'm assigning to @yzhwang.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @yzhwang: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @yzhwang: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @yzhwang: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @yzhwang: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by jart** [comment]
This appears to be a duplicate of #15643. Please follow that issue. If you feel this issue should be reopened, please provide new evidence and I'll do that.

**Comment by jart** [comment]
This appears to be a duplicate of #5592. Please follow that issue. If you feel this issue should be reopened, please provide new evidence and I'll do that.
